File: apps/web/src/lib/api/holds.ts
"use client";

/**
 * The ops work list: which accounts are waiting on a human, and on what.
 *
 * `GET /v1/admin/compliance/holds` (`org:read`, admin realm) shipped in with no screen —
 * two R-11 gates were blocking tenants and the only way to see who was blocked was a
 * curl. `apps/api/admin/holds.py` argues the read: nothing in RLS was widened, the row
 * is composed from the SAME blockers that refuse the client's dial and launch, and it
 * carries accounts rather than people.
 *
 * What this module holds is the console's half of that: the operator vocabulary for
 * each rule, WHICH SCREEN CLEARS IT, and how long an account has been waiting.
 *
 * Three properties of the API kept rather than smoothed over:
 *
 * - **`holds` is a list, not a status.** An account can be held by both gates at once
 *   and the API says so instead of picking a winner, so every row here renders every
 *   rule and offers every remedy. Collapsing them would send an operator to clear KYC
 *   and leave the account still held.
 * - **The rules are the GATES' names, and they arrive as plain strings.** They are the
 *   same words the client's own screen and the launch preview use, so an operator and a
 *   client on the phone name one condition identically. A rule this build has never
 *   heard of is rendered as itself and routed to the account, never dropped: an
 *   unrecognised hold is still a held account.
 * - **Hard rule 6 is a property of the payload, and this module adds nothing to it.**
 *   No phone number, no document reference, no signatory, no reviewer prose — the
 *   reasons the blockers return are dropped server-side on purpose, because
 *   `first_campaign_rejected_reason` interpolates an operator's free text. Nothing here
 *   fetches any of it back; everything identifying stays one click away on the
 *   account's own screen, behind the permission that opens it.
 */

import type { NoticeTone } from "@/components/ui";

import type { components } from "./schema";

type Schemas = components["schemas"];

/** One line of the work list. `holds` is the gates' own rule names. */
export type HeldTenant = Schemas["HeldTenantOut"];

/** One path string for the one endpoint. */
export const HOLDS_PATH = "/v1/admin/compliance/holds";

export interface HoldRule {
  /** What an operator calls this hold. */
  label: string;
  /** What is actually blocked while it stands — the triage question. */
  blocks: string;
  /** The action that clears it, in the operator's words. */
  remedy: string;
  /** The screen that records that action, relative to a tenant id. */
  screen: (tenantId: string) => string;
  /** Wording on the link to that screen. */
  cta: string;
  tone: NoticeTone;
}

/**
 * The four rule names the two gates emit (`compliance/service.py`), each with the
 * screen that clears it.
 *
 * A `Record<string, …>` rather than a `Record<Rule, …>`, because `holds` is `string[]`
 * on the wire — the API deliberately did not narrow it to a Literal, since the set grows
 * whenever a gate does. `holdRule()` below is the only reader, and it fails visible
 * rather than closed: an unknown rule keeps the account on the list and says the console
 * does not know what clears it, which is the honest answer and the one that gets fixed.
 *
 * `kyc_missing` and `kyc_not_verified` are two rules and one screen — they are separate
 * facts (nothing filed at all, versus filed and not cleared) with the same remedy, and
 * `kyc_not_verified_reason` exists in the API for exactly that distinction. The list
 * keeps both names because "we have never heard from them" and "we owe them a review"
 * are different work.
 */
export const HOLD_RULES: Record<string, HoldRule> = {
  kyc_missing: {
    label: "Identity not filed",
    blocks: "All outbound calling, and a number purchase on any tier.",
    remedy: "Chase the business's registration details, then record the verification.",
    screen: (tenantId) => `/admin/tenants/${tenantId}/kyc`,
    cta: "Identity (KYC)",
    tone: "stop",
  },
  kyc_not_verified: {
    label: "Identity not verified",
    blocks: "All outbound calling, and a number purchase on any tier.",
    remedy: "Something is on file and is not cleared — check it and record the outcome.",
    screen: (tenantId) => `/admin/tenants/${tenantId}/kyc`,
    cta: "Identity (KYC)",
    tone: "warn",
  },
  first_campaign_review_pending: {
    label: "First campaign not reviewed",
    blocks: "Every campaign on the account. Inbound answering is unaffected.",
    remedy: "Read the list, the script and the disclosure line, then release the account.",
    screen: (tenantId) => `/admin/tenants/${tenantId}/first-campaign-review`,
    cta: "Review & release",
    tone: "warn",
  },
  first_campaign_review_rejected: {
    label: "First campaign refused",
    blocks: "Every campaign on the account. Inbound answering is unaffected.",
    remedy:
      "A reviewer refused this account. It stays held until the client fixes what was " +
      "named and someone looks again.",
    screen: (tenantId) => `/admin/tenants/${tenantId}/first-campaign-review`,
    cta: "Review & release",
    tone: "stop",
  },
};

/**
 * The copy for a rule, or `null` when this build cannot name it.
 *
 * `Object.hasOwn`, never `rule in HOLD_RULES`: `rule` is an arbitrary string off the
 * wire and `in` walks the prototype chain, so `holdRule("constructor")` handed back
 * `Object` itself typed as a `HoldRule` — and the queue's very next line is
 * `copy.screen(tenantId)`, which is a TypeError that takes the whole ops screen down.
 * The contract here is fail-VISIBLE (an unknown rule keeps the row and says so); a
 * blank screen is the one way to fail that an operator cannot see past.
 * Covered by tests/holds.test.ts.
 */
export function holdRule(rule: string): HoldRule | null {
  return Object.hasOwn(HOLD_RULES, rule) ? HOLD_RULES[rule] : null;
}

/**
 * How long this account has been waiting, in whole hours.
 *
 * `signed_up_at` is on the row precisely so the wait is visible: both gates are "since
 * you signed up, nobody has looked", and the API deliberately did not send a per-gate
 * timestamp, because a KYC record filed later does not restart the account's wait.
 *
 * Clamped at zero. A signup in the future is clock skew between the operator's laptop
 * and the database, not a negative wait, and "in 2 hours" in a waiting column is the
 * kind of nonsense that makes a whole screen untrustworthy.
 */
export function hoursWaiting(signedUpAt: string, now: number = Date.now()): number {
  const started = new Date(signedUpAt).getTime();
  if (Number.isNaN(started)) return 0;
  return Math.max(0, Math.floor((now - started) / 3_600_000));
}

/**
 * "3 days" / "5 hours" — the wait as a phrase, so nobody subtracts dates by hand.
 *
 * `Intl.RelativeTimeFormat` with `numeric: "always"`, which is what turns -3 days into
 * "3 days ago" rather than the conversational "yesterday" that `"auto"` produces
 * (developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/RelativeTimeFormat/format).
 * A queue is scanned, not read: "yesterday" and "2 days ago" next to each other cannot
 * be compared at a glance, and comparing them IS the triage. The "ago" is trimmed here
 * because the column already says what the number measures.
 */
const RELATIVE = new Intl.RelativeTimeFormat("en-IN", { numeric: "always" });

export function waitedFor(signedUpAt: string, now: number = Date.now()): string {
  const hours = hoursWaiting(signedUpAt, now);
  // Below the first whole hour the formatter would render `-0` as "in 0 hours", which is
  // both wrong-tensed and the only cell on the screen pointing at the future. A signup
  // this minute has waited no measurable time, and saying so is the honest reading.
  if (hours === 0) return "under an hour";
  const [value, unit]: [number, Intl.RelativeTimeFormatUnit] =
    hours < 24 ? [hours, "hour"] : [Math.floor(hours / 24), "day"];
  return RELATIVE.format(-value, unit).replace(/\s*ago$/, "");
}

/**
 * How alarming this wait is. THREE BANDS, and they are OURS, not a published SLA.
 *
 * No document in `docs/` sets a turnaround for either gate, so this does not pretend to
 * measure one. What it measures is the shape of the work: both gates are a person
 * reading a page, so an account still waiting on the second day is not busy, it is
 * stuck, and one waiting past a week is the account `holds.py` describes as having
 * "quietly stopped trying" — a self-serve signup that cannot dial and hears nothing does
 * not complain, it churns. The bands exist to make that visible at a glance; the
 * numbers are a triage heuristic and should move the day an SLA is actually agreed.
 */
export const WAIT_WARN_HOURS = 48;
export const WAIT_BREACH_HOURS = 24 * 7;

export function waitBand(signedUpAt: string, now: number = Date.now()): NoticeTone {
  const hours = hoursWaiting(signedUpAt, now);
  if (hours >= WAIT_BREACH_HOURS) return "stop";
  if (hours >= WAIT_WARN_HOURS) return "warn";
  return "neutral";
}

/**
 * The queue's cache key, named here and imported by both the read and every write that
 * can empty a row off it.
 *
 * Clearing either gate removes an account from the list, so the KYC write and the
 * first-campaign decision both have to invalidate this — and a key spelled out at three
 * call sites is a key that eventually differs at one of them, leaving an operator
 * looking at a row they just cleared.
 */
export const HOLDS_QUERY_KEY = ["admin", "holds"] as const;

File: apps/api/admin/holds.py
"""The hold queue — which accounts are waiting on a human, and on what.

Two R-11 mitigations block a tenant until a person at Calevate acts: subscriber KYC
(`compliance/kyc.py`) and the first-campaign review (`compliance/first_campaign.py`).
Both shipped without the other half of the control. An operator found out an account was
held when the client emailed to ask why nothing worked, which makes a mitigation depend
on the client complaining — that is a support queue, not a control, and it is worst for
exactly the accounts that never complain and simply churn.

WHY THIS IS NOT A CROSS-TENANT QUERY (hard rule 1, decided rather than deferred)
-------------------------------------------------------------------------------
Listing held accounts is a cross-tenant read from the admin realm, and `app.admin`
widens `USING` on `organizations` and nothing else (migration b57e2f9c4a13). Three
shapes were available:

1. **Widen `kyc_records` and `first_campaign_reviews` for `app.admin`** the way
   `organizations` was widened, and answer the question in one SQL statement. Rejected.
   A policy is table-scoped, not column-scoped: widening `kyc_records` hands every
   future query on an admin session the signatory's name, the document reference and
   the rejection prose — a permanent grant, to answer a question that needs none of
   them. It also forces the "is this tenant waiting" condition to be re-expressed in
   SQL beside the Python one the gates ask, and two spellings of one compliance
   condition is the drift both modules are written to prevent. b57e2f9c4a13's own
   argument is "the narrowest fix that works"; here a narrower one exists.
2. **Surface the state on the tenant DIRECTORY instead.** This is what
   `first_campaign_routes.py` said the answer should be, and it is half of what this
   module does: `admin.service.tenant_overview` now carries `holds` on every row, from
   the function below, so the screen an operator already reads shows the flag.
3. **The queue itself: the directory, then the tenant's own RLS session.** Which is
   shape 2's mechanism, and the mechanism `tenant_overview` has always used to count a
   client's calls: enumerate tenants under `app.admin`, then enter each tenant with
   `tenant_session()` and ask the ordinary gate. Nothing is widened, no policy changes,
   and the answer comes from the same predicate that refuses the client's dial.

The cost is N+1 by construction — the same trade `tenant_overview` documents, bounded
here by the tier pre-filter below, and payable at M1 scale (one to a few dozen
self-serve accounts) against widening RLS for a work list. A materialized queue is the
answer if it ever gets long enough to notice; it is not the answer today.

WHAT IS ON IT, AND WHAT DELIBERATELY IS NOT (hard rule 6)
---------------------------------------------------------
Accounts, not people. The row carries the tenant, its motion, when it signed up and
which RULES hold it — the same rule names the launch preview and the client's own
screen use, so one condition is never explained in two vocabularies. It carries no
phone number, no document reference, no signatory and no reviewer's note: the reasons
the blockers return are dropped here on purpose, because
`first_campaign_rejected_reason` interpolates an operator's free text and free text can
carry anything into the widest-read list in the console. Everything identifying stays
one click away on the account's own screen, behind the permission that opens it.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from apps.api.compliance.service import (
    SELF_SERVE_TIERS,
    first_campaign_hold_blocker,
    kyc_blocker,
)
from apps.api.db.session import tenant_session


@dataclass(frozen=True, slots=True)
class TenantHolds:
    """Which human-action gates are holding this tenant, if any."""

    # The gates' own rule names, in the order they are asked. Empty = nobody is waiting.
    rules: tuple[str, ...]

    @property
    def held(self) -> bool:
        return bool(self.rules)


#: "Nobody is waiting on us for this account", as a value rather than as a literal.
#:
#: It exists because a caller that can decide the answer WITHOUT asking still has to
#: produce the same type `read_tenant_holds` produces — `tenant_overview` skips the read
#: entirely for a tier no blocker applies to (D-218). Spelling `TenantHolds(rules=())` at
#: that call site would be a second place that knows what an empty answer looks like, and
#: the first thing to drift if this dataclass ever grows a field.
NO_HOLDS = TenantHolds(rules=())


async def read_tenant_holds(session: AsyncSession, *, tenant_id: UUID) -> TenantHolds:
    """THE "is this tenant waiting on us" predicate, on the caller's RLS-scoped session.

    Composed of the gates themselves rather than re-derived: `kyc_blocker` is what
    refuses the dial and `first_campaign_hold_blocker` is what refuses the launch, so a
    queue built from them cannot tell an operator an account is clear while the client
    is staring at a refusal — or leave an account off the list because the queue's own
    copy of the tier line drifted from `SELF_SERVE_TIERS`. Both already fail CLOSED to
    held (absence of a record IS the held state) and both already answer `None` for the
    tiers the controls do not apply to, so this function adds no judgement of its own.

    The `reason` half of each blocker is discarded on purpose — see the module
    docstring, hard rule 6.
    """
    blockers = (
        await kyc_blocker(session, tenant_id=tenant_id),
        await first_campaign_hold_blocker(session, tenant_id=tenant_id),
    )
    # `blocker[0]` is the rule name; `blocker[1]` is the client-facing reason, which
    # stops here.
    return TenantHolds(rules=tuple(blocker[0] for blocker in blockers if blocker is not None))


@dataclass(frozen=True, slots=True)
class HeldTenant:
    """One line of the ops work list."""

    tenant_id: UUID
    name: str
    slug: str
    plan_tier: str
    # When the account arrived. The wait started then: both gates are "since you signed
    # up, nobody has looked", and a per-gate timestamp would be a second answer to one
    # question (a KYC record filed later does not restart the account's wait).
    signed_up_at: datetime
    rules: tuple[str, ...]


# Only these tiers can be held: both blockers return None for every other motion. This
# is the same constant the gates draw their line with, imported rather than repeated, so
# it is a pre-FILTER on the candidate set and never a second copy of the RULE — the
# blocker below still decides, and a tenant this filter let through is dropped by it.
_DIRECTORY = (
    "SELECT id, name, slug, plan_tier, created_at FROM organizations "
    "WHERE deleted_at IS NULL AND plan_tier = ANY(:tiers) "
    "ORDER BY created_at"
)


async def held_tenants(directory: AsyncSession) -> list[HeldTenant]:
    """Every account waiting on a human, oldest signup first.

    `directory` must be an `admin_session()` — the only session that can enumerate
    tenants (b57e2f9c4a13). Each tenant is then ENTERED with its own GUC, so the
    compliance rows are read under ordinary RLS and this function holds no cross-tenant
    view of either table at any point.

    Oldest first because that is the triage order: the account nobody has looked at for
    three weeks is the one that has quietly stopped trying.
    """
    rows = (await directory.execute(text(_DIRECTORY), {"tiers": list(SELF_SERVE_TIERS)})).all()

    queue: list[HeldTenant] = []
    for org in rows:
        tenant_id = UUID(str(org[0]))
        async with tenant_session(tenant_id) as scoped:
            holds = await read_tenant_holds(scoped, tenant_id=tenant_id)
        if not holds.held:
            continue
        queue.append(
            HeldTenant(
                tenant_id=tenant_id,
                name=str(org[1]),
                slug=str(org[2]),
                plan_tier=str(org[3]),
                signed_up_at=org[4],
                rules=holds.rules,
            )
        )
    return queue


__all__ = [
    "NO_HOLDS",
    "HeldTenant",
    "TenantHolds",
    "held_tenants",
    "read_tenant_holds",
]
